### PR TITLE
Prevent UI crash on FileNotFound errors with Active Storage

### PR DESCRIPTION
**Pull request by cpfergus1**
**Description**
In a previous PR #4026, a fix was implemented to prevent an application from crashing by capturing an exception and providing a replacement image to prevent a non-operable UI. In issue #4095 it was found that the implemented fix did not capture all failure points that may be caused by a corrupt or FileNotFound issue.

This commit adds the ability for the application to catch missing images with Active Storage during an API call should a file become corrupt or deleted without removing the record. This compliments PR #4026 which prevented the same error on the backend. The change will return a json image package with attributes equal to nil and noimage replacements if an
image cannot be sourced. This should be back-ported to 3.0 if cleared for merging. 

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [ ] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)
- [ ] I have attached screenshots to this PR for visual changes (if needed)

### CODE REVIEW

File: api/spec/requests/spree/api/images_controller_spec.rb
Comment (kennyadsl) [inline api/spec/requests/spree/api/images_controller_spec.rb:78]: `It` is not required in the example text

File: core/app/models/concerns/spree/active_storage_adapter/attachment.rb
Comment (kennyadsl) [inline core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: @elia @filippoliverani any thoughts on this? I can't think of a better way to fix this problem but maybe we already faced something similar or have other ideas?

File: api/spec/requests/spree/api/images_controller_spec.rb
Comment (cpfergus1) [reply api/spec/requests/spree/api/images_controller_spec.rb:78]: done!

**Review by spaghetticode — COMMENTED** [review]
@cpfergus1 thanks for working on this issue! 

Unless we want to patch [ActiveStorage](https://github.com/rails/rails/blob/581aac754c6cbe9f3f4de396bc8ef3728e56f4f4/activestorage/lib/active_storage/analyzer/image_analyzer/image_magick.rb#L12) directly, this change looks reasonable to me. 
Also, having `nil` instead of `0` for `height` and `width` seems to be the [right way](https://github.com/rails/rails/blob/581aac754c6cbe9f3f4de396bc8ef3728e56f4f4/activestorage/test/analyzer/image_analyzer/image_magick_test.rb#L44) to me. 
On the other hand, not really sure we want `nil` for `identified` and `analyzed`, for example the latter seems to be always set to `true` by [rails](https://github.com/rails/rails/blob/581aac754c6cbe9f3f4de396bc8ef3728e56f4f4/activestorage/app/models/active_storage/blob/analyzable.rb#L51), after extracting the image metadata.
Do you have any opinions on this?

File: core/app/models/concerns/spree/active_storage_adapter/attachment.rb
Comment (waiting-for-dev) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: @elia @filippoliverani pinging you again to move it forward :slightly_smiling_face:

Comment (filippoliverani) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: The solution makes sense to me, but I'd prefer to let the error bubble up and handle it where the attachment is used. That's a breaking error, and different controllers or views might want to handle it in different ways, returning a null object might be misleading 🤔

Comment (kennyadsl) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: I'm not sure about this solution, but please correct me if I'm wrong. If there are 20 images on the page and only one of them is unavailable because of this problem, bubbling up the exception to the controller/view layer we are not able to display the other 19 images. We can only show an error to the user.

Comment (waiting-for-dev) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: Don't you think that could be the desired behavior sometimes? Depending on the model, you might want to catch that issue instead of silencing it. For instance, as an administrator, you want to notice when you upload an image for a product, but it's corrupted.

But I made an error when I said, "we could manage it at the controller layer with a configurable error handler". We should just make returning a null object an optional behavior. If someone asks for it, we can do more work and make it configurable on a model basis.

Comment (kennyadsl) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: My opinion here is that, if this failure will prevent people to buy things from the store, I think that showing a broken image is a better solution. So maybe storefront and APIs should never break because of this. 

On the other side, the admin should notice that there's something wrong, but if we just show the error, what can they do to fix it? I'm afraid that they need to ask some devs to manually destroy the record and try to reupload the image, and probably it's better to keep the admin working until that happens.

Comment (waiting-for-dev) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: > My opinion here is that, if this failure will prevent people to buy things from the store, I think that showing a broken image is a better solution. So maybe storefront and APIs should never break because of this.

What I meant is: A user wants to buy a shirt, but, as the image is not rendered, they don't buy it. As no error is logged, the admin will never know about the problem unless they revisit the product page for some reason.

> On the other side, the admin should notice that there's something wrong, but if we just show the error, what can they do to fix it?

That's an excellent point. Admin shouldn't break. We need different behavior depending on the component.

A good middle-way solution could be returning a nil image but showing a warning so that it's logged and can be detected. What do you think?

Comment (kennyadsl) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: Yes, or delete the corrupted image and continue with the execution. Could be dangerous, I don't know the exact implications of this proposal but if I got it correctly, admin will need to reupload the image again in this case because it's not saved in the storage provider. Why keep an instance in our system for that image the actual image is nowhere?

Comment (waiting-for-dev) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: Not sure, but maybe the stored metadata could help debug?

Comment (kennyadsl) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: Probably, I don't even know if this is reproducible, to be honest.

Comment (waiting-for-dev) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: @cpfergus1, when you have a moment, could you please add a warning when the image is not found? 🙏

Comment (waiting-for-dev) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: Hey, @cpfergus1, pinging you again in case you have time to put the little extra effort needed to complete this 💪

Comment (cpfergus1) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb:58]: Sorry, I missed the request, I will pick it back up!

**Comment by cpfergus1** [comment]
> @cpfergus1 thanks for working on this issue!
> 
> Unless we want to patch [ActiveStorage](https://github.com/rails/rails/blob/581aac754c6cbe9f3f4de396bc8ef3728e56f4f4/activestorage/lib/active_storage/analyzer/image_analyzer/image_magick.rb#L12) directly, this change looks reasonable to me. Also, having `nil` instead of `0` for `height` and `width` seems to be the [right way](https://github.com/rails/rails/blob/581aac754c6cbe9f3f4de396bc8ef3728e56f4f4/activestorage/test/analyzer/image_analyzer/image_magick_test.rb#L44) to me. On the other hand, not really sure we want `nil` for `identified` and `analyzed`, for example the latter seems to be always set to `true` by [rails](https://github.com/rails/rails/blob/581aac754c6cbe9f3f4de396bc8ef3728e56f4f4/activestorage/app/models/active_storage/blob/analyzable.rb#L51), after extracting the image metadata. Do you have any opinion on this?

My thoughts on this are that if the object cannot be found, it cannot be analyzed or identified since it doesn't exist. This behavior can be witnessed on [images_controller_spec.rb](api/spec/requests/spree/api/images_controller_spec.rb) when examining the image blob. If you attempt to run analyze on it - it will fail due to the image not being found, so the object has not been analyzed.

File: core/app/models/concerns/spree/active_storage_adapter/attachment.rb
Comment (cpfergus1) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb]: I have added a point to capture this error in the logs, did we still want this to bubble up to the controller?

Comment (waiting-for-dev) [reply core/app/models/concerns/spree/active_storage_adapter/attachment.rb]: Thanks, @cpfergus1! No, I think it's better not to bubble up to the controller at this point.

**Review by waiting-for-dev — APPROVED** [review]
Many thanks for your continuous work on it, @cpfergus1! A last minor suggestion, it would be nice to add the id of the attachment in the warning message, so that users can't more easily find it.

It'd be also better to rebase from master before merging :pray:

**Review by waiting-for-dev — APPROVED** [review]

**Comment by waiting-for-dev** [comment]
Thanks, @cpfergus1. We only need another review. Pinging @kennyadsl as he also looked into it.

**Review by kennyadsl — APPROVED** [review]

**Pull request merged**
